src/module.c

```diff
@@ -1109,213 +1080,667 @@ int RM_CreateSubcommand(RedisModuleCommand *parent, const char *name, RedisModul
+ *     * `REDISMODULE_ARG_TYPE_UNIX_TIME`: Integer, but Unix timestamp.
+ *     * `REDISMODULE_ARG_TYPE_PURE_TOKEN`: Argument doesn't have a placeholder.
+ *       It's just a binary token. (TODO: Add example.)
+ *     * `REDISMODULE_ARG_TYPE_ONEOF`: Used when user can choose only one of a
+ *       few sub-arguments. Requires `subargs`. (TODO: Add example.)
+ *     * `REDISMODULE_ARG_TYPE_BLOCK`: Used when one wants to group together
+ *       several sub-arguments, usually to apply something on all of them (like
+ *       making the entire group "optional"). Requires `subargs`. (TODO: Add
+ *       example.)
+ *
+ *     Explanation of the command argument flags:
+ *
+ *     * `REDISMODULE_CMD_ARG_OPTIONAL`: The argument is optional (like GET in
+ *       the SET command).
+ *     * `REDISMODULE_CMD_ARG_MULTIPLE`: The argument may repeat itself (like
+ *       key in DEL).
+ *     * `REDISMODULE_CMD_ARG_MULTIPLE_TOKEN`: The argument may repeat itself,
+ *       and so does its token (like `GET pattern` in SORT).
+ *
+ * On success REDISMODULE_OK is returned. On error REDISMODULE_ERR is returned
+ * and `errno` is set to EINVAL if invalid info was provided or EEXIST if info
+ * has already been set. */
+int RM_SetCommandInfo(RedisModuleCommand *command, const RedisModuleCommandInfo *info) {
+    if (!moduleValidateCommandInfo(info)) {
+        errno = EINVAL;
+        return REDISMODULE_ERR;
+    }
 
-/* Returns 1 if `cmd` is a command of the module `modulename`. 0 otherwise. */
-int moduleIsModuleCommand(void *module_handle, struct redisCommand *cmd) {
-    if (cmd->proc != RedisModuleCommandDispatcher)
-        return 0;
-    if (module_handle == NULL)
-        return 0;
-    RedisModuleCommand *cp = (void*)(unsigned long)cmd->getkeys_proc;
-    return (cp->module == module_handle);
-}
+    struct redisCommand *cmd = command->rediscmd;
 
-void extendKeySpecsIfNeeded(struct redisCommand *cmd) {
-    /* We extend even if key_specs_num == key_specs_max because
-     * this function is called prior to adding a new spec */
-    if (cmd->key_specs_num < cmd->key_specs_max)
-        return;
+    /* Check if any info has already been set. Overwriting info involves freeing
+     * the old info, which is not implemented. */
+    if (cmd->summary || cmd->complexity || cmd->since || cmd->history ||
+        cmd->tips || cmd->args ||
+        !(cmd->key_specs_num == 0 ||
+          /* Allow key spec populated from legacy (first,last,step) to exist. */
+          (cmd->key_specs_num == 1 && cmd->key_specs == cmd->key_specs_static &&
+           cmd->key_specs[0].begin_search_type == KSPEC_BS_INDEX &&
+           cmd->key_specs[0].find_keys_type == KSPEC_FK_RANGE))) {
+        errno = EEXIST;
+        return REDISMODULE_ERR;
+    }
 
-    cmd->key_specs_max++;
+    if (info->summary) cmd->summary = zstrdup(info->summary);
+    if (info->complexity) cmd->complexity = zstrdup(info->complexity);
+    if (info->since) cmd->since = zstrdup(info->since);
```

Comment:
So in practice we can require that all `const char *` are literals (or at least static) but we'll copy everything else? That's good because ATM we nees to cast the strings to `char *` to free them.

Reply:
i think that'll be a good approach (don't dup these strings).
@yossigo @MeirShpilraien @guybe7?

Reply:
i also don't like that we cast `const` stuff just to release it but I think we should still copy due to Yossi's points

Reply:
Keeping dup for now. Resolved?

Reply:
yes. resolving.